agent-inject/agent/agent.go

```diff
@@ -309,10 +317,44 @@ func (a *Agent) Patch() ([]byte, error) {
 		if err != nil {
 			return patches, err
 		}
-		a.Patches = append(a.Patches, addContainers(
-			a.Pod.Spec.InitContainers,
-			[]corev1.Container{container},
-			"/spec/initContainers")...)
+
+		containers := a.Pod.Spec.InitContainers
+
+		// Init Containers run sequentially in Kubernetes and sometimes the order in
+		// which they run matters.  This reorders the init containers to put the agent first.
+		// For example, if an init container needed Vault secrets to work, the agent would need
+		// to run first.
+		if a.InitFirst {
+
+			// Remove all init containers from the document so we can re-add them after the agent.
+			if len(a.Pod.Spec.InitContainers) != 0 {
+				a.Patches = append(a.Patches, removeContainers("/spec/initContainers")...)
+			}
+
+			containers = []corev1.Container{container}
+			containers = append(containers, a.Pod.Spec.InitContainers...)
+
+			a.Patches = append(a.Patches, addContainers(
+				[]corev1.Container{},
+				containers,
+				"/spec/initContainers")...)
```

Comment:
I'm curious why `containers` is created with all the containers before this, instead of doing something like:

```suggestion
			a.Patches = append(a.Patches, addContainers(
				[]corev1.Container{container},
				a.Pod.Spec.InitContainers,
				"/spec/initContainers")...)
```

Reply:
This doesn't work due to the way `addContainers` is implemented.  The first arg is just the target, IE how many are existing currently.  Since we removed all containers before, we send an empty slice.